server/src/actors/outcome.rs

```diff
@@ -0,0 +1,416 @@
+
+/// This is the implementation that uses kafka queues and does stuff
+#[cfg(feature = "processing")]
+mod real_implementation {
+    use super::*;
+
+    use futures::Future;
+    use rdkafka::{
+        error::KafkaError,
+        producer::{FutureProducer, FutureRecord},
+        ClientConfig,
+    };
+    use serde_json::Error as SerdeSerializationError;
+
+    use crate::actors::controller::{Controller, Shutdown, Subscribe, TimeoutError};
+    use crate::service::ServerErrorKind;
+    use crate::utils::{SyncFuture, SyncHandle};
+    use failure::{Fail, ResultExt};
+    use semaphore_common::tryf;
+    use semaphore_common::{metric, KafkaTopic, Uuid};
+
+    #[derive(Fail, Debug)]
+    pub enum OutcomeError {
+        #[fail(display = "kafka message send canceled")]
+        Canceled,
+        #[fail(display = "failed to send kafka message")]
+        SendFailed(KafkaError),
+        #[fail(display = "shutdown timer expired")]
+        Shutdown,
+        #[fail(display = "json serialization error")]
+        SerializationError(SerdeSerializationError),
+        #[fail(display = "internal error")]
+        InternalError,
+    }
+
+    pub struct OutcomeProducer {
+        config: Arc<Config>,
+        shutdown: SyncHandle,
+        producer: Option<FutureProducer>,
+        is_enabled: bool,
+    }
+
+    impl OutcomeProducer {
+        pub fn create(config: Arc<Config>) -> Result<Self, ServerError> {
+            let processing_enabled = config.processing_enabled();
+            let future_producer = if processing_enabled {
+                let mut client_config = ClientConfig::new();
+                for config_p in config.kafka_config() {
+                    client_config.set(config_p.name.as_str(), config_p.value.as_str());
+                }
+                let future_producer = client_config
+                    .create()
+                    .context(ServerErrorKind::KafkaError)?;
+                Some(future_producer)
+            } else {
+                None
+            };
+
+            Ok(OutcomeProducer {
+                config,
+                shutdown: SyncHandle::new(),
+                producer: future_producer,
+                is_enabled: processing_enabled,
+            })
+        }
+    }
+
+    impl Actor for OutcomeProducer {
+        type Context = Context<Self>;
+
+        fn started(&mut self, context: &mut Self::Context) {
+            log::info!("OutcomeProducer started.");
+            //register with the controller for shutdown notifications
+            Controller::from_registry().do_send(Subscribe(context.address().recipient()));
+        }
+
+        fn stopped(&mut self, _ctx: &mut Self::Context) {
+            log::info!("OutcomeProducer stopped.");
+        }
+    }
+
+    impl Handler<OutcomeMessage> for OutcomeProducer {
+        type Result = ResponseFuture<(), OutcomeError>;
+
+        fn handle(&mut self, message: OutcomeMessage, _ctx: &mut Self::Context) -> Self::Result {
```

Comment:
_nit_: What if we emit a debug log here? This could be nice to debug Relay - even if processing is disabled - to see where events go. We can even do this in the noop implementation, because it might be valuable information.

Reply:
ok, done